### PR TITLE
bugfix(node): make WriteMode optional when specifying embeddings

### LINKED ISSUE #331: Write mode should be optional when specifying an embedding
We should be able to specify embeddings without having to pass along the write mode

```
createTable('table', [ { ... } ], openai)
```

https://lancedb.github.io/lancedb/javascript/interfaces/Connection/#createtable

### PR DESCRIPTION
**Pull request by gsilvestrin**
- Created a WriteOptions to hold WriteMode and any future optional params to createTable
- Implemented user typeguards for WriteOptions
- Move overloaded function declarations to the interface instead of impl

Close #331

### CODE REVIEW

File: node/src/embedding/embedding_function.ts [hunk 1]
```diff
@@ -26,3 +26,9 @@ export interface EmbeddingFunction<T> {
      */
   embed: (data: T[]) => Promise<number[][]>
 }
+
+export function isEmbeddingFunction<T> (value: any): value is EmbeddingFunction<T> {
+  return Object.keys(value).length === 2 &&
```
Comment (eddyxu) [inline node/src/embedding/embedding_function.ts:31]: Not familar with how typescript works. But what if value is an object that do not have `sourceColumn` and `embed` fields, but other fields?

File: node/src/index.ts [hunk 2]
```diff
@@ -97,11 +99,39 @@ export interface Connection {
...
+   *
+   * @param {string} name - The name of the table.
+   * @param data - Non-empty Array of Records to be inserted into the table
+   * @param {EmbeddingFunction} embeddings - An embedding function to use on this table
+   * @param {WriteOptions} options - The write options to use when creating the table.
+   */
+  createTable<T> (name: string, data: Array<Record<string, unknown>>, embeddings: EmbeddingFunction<T>, options: WriteOptions): Promise<Table<T>>
 
+  /**
+   * @ignore
```
Comment (eddyxu) [inline node/src/index.ts:133]: what does ignore mean? Do we still want this function?

File: node/src/index.ts [hunk 3]
```diff
@@ -234,32 +264,19 @@ export class LocalConnection implements Connection {
...
-   * @param data Non-empty Array of Records to be inserted into the Table
-   * @param mode The write mode to use when creating the table.
-   * @param embeddings An embedding function to use on this Table
-   */
-  async createTable<T> (name: string, data: Array<Record<string, unknown>>, mode: WriteMode, embeddings: EmbeddingFunction<T>): Promise<Table<T>>
-  async createTable<T> (name: string, data: Array<Record<string, unknown>>, mode: WriteMode, embeddings?: EmbeddingFunction<T>): Promise<Table<T>>
-  async createTable<T> (name: string, data: Array<Record<string, unknown>>, mode: WriteMode, embeddings?: EmbeddingFunction<T>): Promise<Table<T>> {
-    if (mode === undefined) {
-      mode = WriteMode.Create
+  async createTable<T> (name: string, data: Array<Record<string, unknown>>, optsOrEmbedding?: WriteOptions | EmbeddingFunction<T>, opt?: WriteOptions): Promise<Table<T>> {
```
Comment (eddyxu) [inline node/src/index.ts:267]: is this right that `WriteOptions` can be either 3rd or 4th parameter?  should we just use a PartialObject<{ write_options?: WriteOpitons, embeddings: Embeddigs> }>

Or put `embeddings` as a member of WriteOptions?

**Comment by eddyxu** [comment]
Please edit the PR title.

File: node/src/index.ts [hunk 4]
```diff
@@ -459,6 +473,23 @@ export enum WriteMode {
...
+ * Write options when creating a Table.
+ */
+export interface WriteOptions {
+  /** A {@link WriteMode} to use on this operation */
+  writeMode?: WriteMode
+}
+
+export class DefaultWriteOptions implements WriteOptions {
+  writeMode = WriteMode.Create
+}
```
Comment (wjones127) [inline node/src/index.ts:486]: What's the point of "write mode" in create table? Are there any other variants?

**Review by wjones127 — APPROVED** [review]

**Pull request merged**